src/Dscp/Educator/Validate.hs

```diff
@@ -2,34 +2,67 @@
 -- | Validation of private block
 
 module Dscp.Educator.Validate
-       ( validatePrivateBlk
+       ( ValidationFailure (..)
+       , validatePrivateBlk
        ) where
 
 import Universum
 
-import Serokell.Util.Verify (verResToMonadError, verifyGeneric)
+import Data.Text.Buildable (build)
 
 import Dscp.Core (Address (..), SignedSubmission (..), Submission (..),
-                  SubmissionWitness (..))
-import Dscp.Crypto (fromFoldable, getMerkleRoot, hash, verify)
+                  SubmissionSig, SubmissionWitness (..))
+import Dscp.Crypto (Hash, PublicKey, MerkleSignature, fromFoldable, getMerkleRoot, hash, verify)
 import Dscp.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), pbBody, pbbTxs)
 import Dscp.Educator.Serialise ()
 import Dscp.Educator.Txs (PrivateTx (..))
 
+import qualified Text.Show
+import qualified Data.Text.Buildable ()
+
+
+-- | Validation failures
+data ValidationFailure
+    = MerkleSignatureMismatch !(MerkleSignature PrivateTx) !(MerkleSignature PrivateTx)
+      -- ^ Header merkle tree root does not match
+      -- body merkle tree root
+    | TxPublicKeyMismatch !(Hash PublicKey) !(Hash PublicKey)
+      -- ^ Witness public key do not match transaction address
+    | TxSignatureMismatch !PublicKey !SubmissionSig !(Hash Submission)
+      -- ^ Witness signature do not match transaction submission signature
+    deriving (Eq)
+
+instance Show ValidationFailure where
+    show e = toString . pretty $ e
```

Comment:
Eta-reduction: `show = toString . pretty`.
(You can use `hlint` to see such hints automatically)